File: crc16/hash.go
// Calculates the CRC-16 checksum of a byte slice using the specified polynomial.
func crc16(data []byte, polynomial uint16) uint16 {
    const initialCrc uint16 = 0xffff
    crc := initialCrc

    for _, b := range data {
        crc ^= uint16(b) << 8
        for i := 0; i < 8; i++ {
            if crc & 0x8000 != 0 {
                crc = (crc << 1) ^ polynomial
            } else {
                crc <<= 1
            }
        }
    }

    return crc & 0xffff
}

// Calculates the CRC-16 checksum of a byte slice using the standard CRC-16-CCITT polynomial.
func crc16Ccitt(data []byte) uint16 {
    const polynomial uint16 = 0x1021
    return crc16(data, polynomial)
}

File: crc16/hash.rs
/**
 * Calculates the CRC-16 checksum of a byte slice using the specified polynomial.
 *
 * @param data The byte slice to be checksummed.
 * @param polynomial The polynomial to use for the checksum calculation.
 * @returns The 16-bit CRC value as a u16.
 */
 fn crc16(data: &[u8], polynomial: u16) -> u16 {
    let mut crc: u16 = 0xffff;

    for byte in data.iter() {
        crc ^= u16::from(*byte) << 8;
        for _ in 0..8 {
            crc = if crc & 0x8000 != 0 {
                (crc << 1) ^ polynomial
            } else {
                crc << 1
            };
        }
    }

    crc & 0xffff
}

/**
 * Calculates the CRC-16 checksum of a byte slice using the standard CRC-16-CCITT polynomial.
 *
 * @param data The byte slice to be checksummed.
 * @returns The 16-bit CRC value as a u16.
 */
fn crc16_ccitt(data: &[u8]) -> u16 {
    crc16(data, 0x1021)
}
